[4.1] RavenDB-10925 added a surpassed auto index cleanup task

ppekrol opened the pull request.

ppekrol committed:
RavenDB-10925 added a surpassed auto index cleanup task

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1005,6 +1003,61 @@ public IEnumerable<Index> GetIndexes()
+
+        private async Task DeleteSurpassedAutoIndexes()
+        {
+            if (_serverStore.IsLeader() == false)
```
This is wrong. 
What you are doing here is asking for the _cluster_, not the db group.
If we have a db that is not on the leader, it will never run this.
Better to say something like: "dbGroup[0] == this"

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1005,6 +1003,61 @@ public IEnumerable<Index> GetIndexes()
+
+            foreach (var indexName in indexesToRemove)
+            {
+                await TryDeleteIndexIfExists(indexName);
```
Error handling?

ppekrol committed:
RavenDB-10925
- added logging and error handling
- running idle operations only on first member from db group

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1002,15 +1003,25 @@ public IEnumerable<Index> GetIndexes()
+            using (context.OpenReadTransaction())
+                record = _serverStore.Cluster.ReadDatabase(context, _documentDatabase.Name);
+
+            if (record.Topology.Members[0] != _serverStore.NodeTag)
```
@ayende Is there a possibility to have `null` here or 0 members? I need to apply similar fix to v4.0

ayende replied:
No, it can't be null / empty

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1054,17 +1065,22 @@ public void RunIdleOperations()
             }
         }
 
         private async Task HandleUnusedAutoIndexes()
```
I'm not sure how correct this is going to be.  What happens if queries are made on a _different_ node, and not this one?

ayende replied:
Might actually be better to move this entirely to the cluster observer?

ppekrol replied:
Created a separate issue for that: http://issues.hibernatingrhinos.com/issue/RavenDB-11206

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1003,84 @@ public IEnumerable<Index> GetIndexes()
+            await DeleteSurpassedAutoIndexes();
+        }
+
+        private async Task DeleteSurpassedAutoIndexes()
```
Let's run this IFF there have been changes in the indexes since we last run this.

ayende replied:
Can just check the index of the db record to see if it changed

ppekrol replied:
Done

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1003,84 @@ public IEnumerable<Index> GetIndexes()
         }
 
-        private async Task HandleUnusedAutoIndexes()
+        private async Task RunIdleOperationsAsync()
```
Technically speaking, this may actually run on two different machines at the same time, even with the check.
I wonder if we should handle this in some manner? Maybe do the operation optimistically, so send it to the leader and it will only actually run if the db record didn't change?

ppekrol replied:
I have some ideas, let's apply them as a part of http://issues.hibernatingrhinos.com/issue/RavenDB-11206

ppekrol committed:
RavenDB-10925 using database record etag to avoid redundant and unnecessary work in DeleteSurpassedAutoIndexes

ppekrol committed:
RavenDB-10925
- added support for automating index merging
- fixed a bug when we would not extend the GroupBy fields properly

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+            if (record.Topology.Members[0] != _serverStore.NodeTag)
+                return;
+
+            AsyncHelpers.RunSync(() => RunIdleOperationsAsync(etag));
```
Can there be errors from here? Don't we need to handle this?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+
+                    indexesToRemove.Add(kvp.Key);
+                }
+                catch (Exception e)
```
shouldn't we abort if there are errors?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+                try
+                {
+                    await TryDeleteIndexIfExists(indexName);
+                    if (_logger.IsInfoEnabled)
```
Shouldn't we abort if there are errors?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+
+                try
+                {
+                    await CreateIndex(definition);
```
I wonder if it isn't safer to do this one index at a time, instead of all the matches.

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+            if (_lastSurpassedAutoIndexesDatabaseRecordEtag >= databaseRecordEtag)
                 return;
 
+            _lastSurpassedAutoIndexesDatabaseRecordEtag = databaseRecordEtag;
```
If there have been errors, we won't run until the next DatabaseRecord update.
What about moving this to the end?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -1081,9 +1188,17 @@ orderby lastQueryingTime
-                            _logger.Info($"Deleted index '{item.Index.Name}' due to idleness. Age: {age}. Last query: {lastQuery}.");
+                        try
+                        {
+                            await TryDeleteIndexIfExists(item.Index.Name);
```
I think we should do this one index at a time.

ayende reviewed (commented):
We still need to ensure that this doesn't run concurrently, which means that we need to add the original database record etag to the requests.
